channels/offline-channel-js/src/OfflineChannel.ts

```diff
@@ -0,0 +1,614 @@
+                        size += payload.length;
+                        if (size > _maxBatchSize) {
+                            return;
+                        }
+                        if(curEvt.persistence == EventPersistence.Critical) {
+                            criticalCnt ++;
+                        }
+                        idx = index;
+                        payloadArr.push(payload);
+
+                    });
+                    if (!payloadArr.length) {
+                        return;
+                    }
+                    
+                    let sentItems = evts.slice(0, idx + 1);
+                 
+                    _inMemoBatch = _inMemoBatch.createNew(_endpoint, inMemo.getItems().slice(idx + 1));
+                   
+                    let payloadData = _constructPayloadData(payloadArr, criticalCnt);
+                    let callback: OfflineBatchStoreCallback = (res) => {
+                        if (res.state == eBatchStoreStatus.Failure ) {
+                            arrForEach(sentItems, (item) => {
+                                _inMemoBatch.addEvent(item);
+                            });
+                        }
+                    };
+                    if (payloadData) {
+                        let promise = _urlCfg.batchHandler.storeBatch(payloadData, callback, unload);
+                        _queueStorageEvent("storeBatch", promise);
+                    }
+
+                } catch (e) {
+                    // eslint-disable-next-line no-empty
+
+                }
+                return null;
+
+            }
+
+            function _setSendNextTimer() {
+           
+                if(!_sendNextBatchTimer) {
+                    let retryInterval = _retryAt ? Math.max(0, _retryAt - dateNow()) : 0;
+                    let timerValue = Math.max(_maxBatchInterval, retryInterval);
+                    _sendNextBatchTimer = scheduleTimeout(() => {
+                        if (_offlineListener.isOnline()) {
+                            if(!_sender.isCompletelyIdle()) {
+                                _sendNextBatchTimer.refresh();
+
+                            } else {
+                                let callback:  OfflineBatchCallback = (res) => {
+                                    if (res.state !== eBatchSendStatus.Complete) {
+                                        _consecutiveErrors ++;
+                                    }
+                                    _sendNextBatchTimer && _sendNextBatchTimer.refresh();
+                                }
+                                let promise = _urlCfg.batchHandler.sendNextBatch(callback, false, _senderInst);
+                                _queueStorageEvent("sendNextBatch", promise);
+                            }
+                           
+                        } else {
+                            _sendNextBatchTimer.cancel();
+                        }
+                        // if offline, do nothing;
```

Comment:
We should also "stop" (cancel) the timer if we don't have anything in the "next" queue (ie. we need should check the `hasStoredBatch` and only "restart" / "start" the timer if it returns true. Which we would also do after "adding" an in memory batch (because we know the answer will be true)  -- except of course like you have here that we believe we are offline.

Reply:
created task 26681249 for this